### PR TITLE
Calypsoify: Requiring the Calypsoify package in the Jetpack and Mu-wpcom plugins

### PR DESCRIPTION
**Pull request by coder-karen**
Fixes https://github.com/Automattic/vulcan/issues/320

## Proposed changes:

* This PR requires the newly created Calypsoify package in both the Jetpack plugin and the mu-wpcom plugin (but makes sure the class initialization in the mu-wpcom package only happens if a site is Atomic, as that mirrors the current behaviour on Simple sites (the Jetpack module assets are not loaded there)).
* It also modifies the webpack.config.js file (simplified - JS dealt with by the main Jetpack config and in terms of CSS it's just compiling CSS from SCSS), imports the scss file in the js file (in 'src'), and changes how assets are registered by using the Assets package.
* There have been some CSS and JS tweaks to fix things that have broken at some stage in the past - predominantly the option to toggle off the full-screen editor is now hidden if Calypsoify is enabled.
* What has been fixed:
  * Styles related to hiding the full-screen toggle option (via Options menu in post editor) if Calypsoified
* What has been removed:
  * Styles to hide full-screen option from site editor view modes (option no longer exists)
  * Script to force full-screen for iframed site editor (can no longer be anything but that)
  * The scripts relevant to the close button have been removed as it's only relevant for Jetpack sites, and we're removing Calypsoify from self-hosted Jetpack (Atomic sites have a nav-bar in place thanks to the ETK plugin, which overrides that button).
  * The scripts relevant to the revision.php page have been removed as that was only relevant for Simple sites. Instead a change was made directly to admin-plugins to make sure the `mods-gutenberg.js` file there is loading correctly on the revisions page. Diff here: D149239-code
* Additional info:
  * The `calypsoifyGutenberg` variable defined in the main php file includes `closeUrl`, `manageReusableBlocksUrl` and `createNewPostUrl`. `calypsoifyGutenberg` is still needed for the posts navigation on Atomic sites - this is the `closeUrl` in particular, and it is added to the sidebar via the ETK plugin. Removing it here does affect the relevant link on Atomic sites.  The other two variables - `manageReusableBlocksUrl` and `createNewPostUrl'` - appear to be used in `/widgets.wp.com/wpcom-block-editor/calypso.editor.js`. I'm not familiar with this and digging in would take considerably more time and is already out of the scope of this project, so for the sake of not breaking anything else I've left the `calypsoifyGutenberg` variable and associated functions intact.
  * Why don't we get rid of the `admin-plugins` version of Calypsoify? Some of the code is WPcom specific (some in the JS file in admin-plugins). There are also small differences in the main php file so workarounds would need to be added in the mu-wpcom package version if possible. In short we could, but for now I'm opting to keep things as is on Simple. Should there ever be more work done on Calypsoify in the future maybe it would be worthwhile. There are also potential unknowns at this stage and more testing needed if removing the admin-plugins version.

Related diff: D149239-code

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

https://github.com/Automattic/vulcan/issues/320

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

Jetpack self hosted and Atomic: 

* To test changes on a self hosted site and Atomic, apply this PR.
* If applying the PR locally, make sure to build the assets with `jetpack build packages/calypsoify --production`, jetpack build packages/jetpack-mu-wpcom  --production` and `jetpack build plugins/mu-wpcom-plugin  --production`. 
* If using the Jetpack Beta plugin to test, follow the steps in the generated comment below, remember to apply the PR both for the Jetpack plugin and the 'WordPress.com Features' plugin via the Beta tester UI. More info here too on WoA testing: PCYsg-Osp-p2 
* Visit the wp-admin post editor for a post (not using the Classic Editor), and add the `calypsoify=1` param to the URL. Example: `https://example.com/wp-admin/post.php?post=827&action=edit&calypsoify=1`.
* Confirm the Calypso stylesheet / scripts are loaded when the query param is added by checking the page source (see `calypsoify_wpadminmods_js-css` and `calypsoify_wpadminmods_js-js-extra`).
* Click on the 'options' menu, confirm that you see no option to toggle Fullscreen mode off.
* Go back to the post and remove the `&calypsoify=1` from the URL, and refresh. Click on the 'options' menu, and you should see the option to toggle off Fullscreen mode. Click this, and refresh the page to confirm the change sticks.
* Then, Calypsoify the URL and refresh the page again - the editor should now be full-screen, and the option to toggle off full-screen gone from the options menu.
* Open up a revision for the post, and Calypsoify the URL ( for example `/wp-admin/revision.php?revision=122&calypsoify=1`). The revision page should now be full-screen (vs. not without Calypsoification).

Atomic-specific:

- Additionally, on an Atomic test site, Calypsoify a post URL and then click on the site icon (top left). A sidebar should display (this is thanks to the ETK plugin). Hover over 'All posts' and the URL should the WordPress.com posts URL. 
- Do the same but without a Calypsoified post - the URL should be the wp-admin posts URL.

Simple site:

The changes added here should not affect Simple sites. The main thing to check is whether or not additional stylesheets or scripts are being loaded on Simple sites (using the steps to test this PR on Simple from the generated comment below) - checking the page source when Calypsoifying a (non-Classic-editor) post you shouldn't see duplications and assets should be coming from `admin-plugins`.

Note that more realistic way to test Calypsoification is to visit `https://wordpress.com/block-editor/post/` and select a self hosted site.

Fixes related to the revisions page and CSS on Simple (copying the changes made here which also is relevant on Simple)  can be tested here: D149239-code

### CODE REVIEW

File: projects/packages/calypsoify/src/class-jetpack-calypsoify.php [hunk 1]
```diff
@@ -68,11 +69,17 @@ public function setup() {
...
-		wp_style_add_data( 'calypsoify_wpadminmods_css', 'rtl', 'replace' );
-		wp_style_add_data( 'calypsoify_wpadminmods_css', 'suffix', '.min' );
 
-		wp_enqueue_script( 'calypsoify_wpadminmods_js', plugin_dir_url( __FILE__ ) . 'build/mods-gutenberg.js', array( 'jquery' ), self::PACKAGE_VERSION, false );
+		Assets::register_script(
+			'calypsoify_wpadminmods_js',
+			'../dist/index.js',
+			__FILE__,
+			array(
+				'jquery',
```
Comment (anomiex) [inline projects/packages/calypsoify/src/class-jetpack-calypsoify.php]: I think you meant like this here
```suggestion
				'dependencies' => array( 'jquery' ),
```

OTOH, if you were to add an `import jQuery from 'jquery';` to the JS file, then it would show up in the `index.asset.php` file and you could omit specifying it here.

But if you're going to update the script, even better would probably be to replace the `$( 'some selector' )` with `document.querySelector()` or `document.querySelectorAll()`, and then the `.attr()` calls with direct accesses to `.href` and `.target`, so we can avoid requiring jquery at all. 🙂 And probably also replace the uses of `wp.data` with `import { dispatch, select } from '@wordpress/data';`.

File: projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php [hunk 2]
```diff
@@ -97,6 +99,11 @@ public static function load_features() {
 		if ( class_exists( 'Automattic\Jetpack\Scheduled_Updates' ) ) {
 			Scheduled_Updates::init();
 		}
+
+		// Gets autoloaded from the Scheduled_Updates package.
+		if ( class_exists( 'Automattic\Jetpack\Calypsoify\Jetpack_Calypsoify' ) ) {
```
Comment (anomiex) [inline projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php]: This works the same but IMO is more readable.
```suggestion
		if ( class_exists( Jetpack_Calypsoify::class ) ) {
```
Although... when wouldn't the class exist, since the package is being required in composer.json?

Comment (coder-karen) [reply projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php]: Good point, I just followed the previous process, but then I see the one above does not check for the class (`\Marketplace_Products_Updater::init();`). Will change that.

File: projects/packages/calypsoify/src/class-jetpack-calypsoify.php (hunk 1 above)
Comment (coder-karen) [reply projects/packages/calypsoify/src/class-jetpack-calypsoify.php]: Good idea :)

File: projects/packages/calypsoify/src/js/mods-gutenberg.js [hunk 3]
```diff
@@ -1,45 +1,40 @@
...
 
-	var editPostHeaderInception = setInterval( function () {
-		var $closeButton = $( '.edit-post-fullscreen-mode-close__toolbar a' );
-		if ( $closeButton.length < 1 ) {
-			return;
-		}
-		clearInterval( editPostHeaderInception );
+var editPostHeaderInception = setInterval( function () {
+	var closeButton = document.querySelector( '.edit-post-fullscreen-mode-close__toolbar a' );
+	if ( closeButton.length < 1 ) {
```
Comment (anomiex) [inline projects/packages/calypsoify/src/js/mods-gutenberg.js]: `querySelector()` returns a node or null, not anything with a length property.
```suggestion
	if ( ! closeButton ) {
```

File: projects/packages/calypsoify/src/js/mods-gutenberg.js [hunk 4]
```diff
@@ -1,45 +1,40 @@
...
-		var href = $( this ).attr( 'href' );
-		if ( href ) {
-			$( this ).attr( 'href', href.replace( '&classic-editor', '' ) );
-		}
-	} );
+$( 'body.revision-php a' ).each( function () {
+	var href = $( this ).getAttribute( 'href' );
+	if ( href ) {
+		$( this ).setAttribute( 'href', href.replace( '&classic-editor', '' ) );
+	}
```
Comment (anomiex) [inline projects/packages/calypsoify/src/js/mods-gutenberg.js]: Missed the jquery uses down here. Something like this should work.
```suggestion
for ( const node of document.querySelectorAll( 'body.revision-php a' ) ) {
	const href = node.getAttribute( 'href' );
	if ( href ) {
		node.setAttribute( 'href', href.replace( '&classic-editor', '' ) );
	}
```

File: projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php [hunk 5]
```diff
@@ -94,16 +92,12 @@ public static function load_features() {
 
 		// Initializers, if needed.
 		\Marketplace_Products_Updater::init();
+		\Jetpack_Calypsoify::get_instance();
```
Comment (anomiex) [inline projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php]: Either bring back the `use` and drop the `\`, or do `\Automattic\Jetpack\Calypsoify\Jetpack_Calypsoify` here.

File: projects/packages/calypsoify/src/js/mods-gutenberg.js [hunk 6]
```diff
@@ -1,44 +1,40 @@
...
 } );
```
Comment (anomiex) [inline projects/packages/calypsoify/src/js/mods-gutenberg.js]: Oops, this too with the above change.
```suggestion
}
```

File: projects/packages/calypsoify/src/js/mods-gutenberg.js (hunk 3 above)
Comment (coder-karen) [reply projects/packages/calypsoify/src/js/mods-gutenberg.js]: Thanks for these! I've not got a good way to test many of these changes just yet so will make the changes but need to follow up later when I can actually test things.

File: projects/packages/calypsoify/src/js/mods-gutenberg.js [hunk 7]
```diff
@@ -1,44 +1,40 @@
...
+}
 
-		if ( calypsoifyGutenberg.closeUrl ) {
-			$closeButton.attr( 'href', calypsoifyGutenberg.closeUrl );
-			$closeButton.attr( 'target', '_parent' );
-		}
-	} );
+var editPostHeaderInception = setInterval( function () {
+	var closeButton = document.querySelector( '.edit-post-fullscreen-mode-close__toolbar a' );
+	if ( closeButton ) {
```
Comment (anomiex) [inline projects/packages/calypsoify/src/js/mods-gutenberg.js]:
```suggestion
	if ( ! closeButton ) {
```

File: projects/packages/calypsoify/src/js/mods-gutenberg.js (hunk 4 above)
Comment (coder-karen) [reply projects/packages/calypsoify/src/js/mods-gutenberg.js]: This no longer worked anyway as we needed to check if the page had loaded properly first, so has been rewritten.

File: projects/packages/calypsoify/.gitattributes [hunk 8]
```diff
@@ -15,3 +15,6 @@ phpunit.xml.dist  production-exclude
 .phpcs.dir.xml    production-exclude
 tests/**          production-exclude
 .phpcsignore      production-exclude
+
+# Files needed in the production build.
+dist/**         production-include
```
Comment (anomiex) [inline projects/packages/calypsoify/.gitattributes]: Consider moving this up to line 8?

File: projects/packages/calypsoify/src/scss/style-gutenberg.scss [hunk 9]
```diff
@@ -2,19 +2,14 @@
  * Hide Fullscreen option from post-editor
  *
  * The entire menu group is not rendered on smaller screens so we need a media query.
- * Hide the third item of the first group when the screen is big enough.
+ * Hide the fourth item of the first group when the screen is big enough.
```
Comment (anomiex) [inline projects/packages/calypsoify/src/scss/style-gutenberg.scss]:
```suggestion
 * Hide the last item of the first group when the screen is big enough.
```

File: projects/packages/calypsoify/src/js/mods-gutenberg.js [hunk 10]
```diff
@@ -1,44 +1,19 @@
...
-
-		if ( calypsoifyGutenberg.closeUrl ) {
-			$closeButton.attr( 'href', calypsoifyGutenberg.closeUrl );
-			$closeButton.attr( 'target', '_parent' );
-		}
-	} );
-
-	$( 'body.revision-php a' ).each( function () {
-		var href = $( this ).attr( 'href' );
+document.addEventListener( 'DOMContentLoaded', function () {
```
Comment (anomiex) [inline projects/packages/calypsoify/src/js/mods-gutenberg.js]: I wonder whether it'd be better to use [`@wordpress/dom-ready`](https://www.npmjs.com/package/@wordpress/dom-ready) instead.

File: projects/packages/calypsoify/webpack.config.calypsoify.js [hunk 11]
```diff
@@ -1,82 +1,40 @@
...
-			I18nCheckPlugin: false,
-		} ),
-		//Delete the dummy JS files Webpack would otherwise create.
-		new RemoveAssetWebpackPlugin( {
-			assets: name => name === 'style-gutenberg.js',
-		} ),
-	],
-};
-
+//
```
Comment (anomiex) [inline projects/packages/calypsoify/webpack.config.calypsoify.js]: Why the empty comment here, rather than just an empty line?

**Review by anomiex — COMMENTED** [review]
Looking reasonable. A few nitpicks inline. I haven't tested it though.

File: projects/packages/calypsoify/webpack.config.calypsoify.js (hunk 11 above)
Comment (coder-karen) [reply projects/packages/calypsoify/webpack.config.calypsoify.js]: No reason, was left behind, removed now.

File: projects/packages/calypsoify/src/js/mods-gutenberg.js (hunk 10 above)
Comment (coder-karen) [reply projects/packages/calypsoify/src/js/mods-gutenberg.js]: I've now removed all of this. I've worked on a separate diff to make sure the admin-plugins version of Calypsoify works properly instead.

File: projects/packages/calypsoify/.gitattributes (hunk 8 above)
Comment (coder-karen) [reply projects/packages/calypsoify/.gitattributes]: Good point!

**Review by fgiannar — COMMENTED** [review]
On Simple:
- Same styles seem to be coming from 2 different places:
1. `wp-content/admin-plugins/calypsoify/style-gutenberg.css?m=1676281014i&ver=0.2.1`
2. `wp-content/mu-plugins/jetpack-mu-wpcom-plugin/moon/vendor/automattic/jetpack-calypsoify/dist/index.css?m=1716281754i&minify=false&ver=b40fe4e9b3851a31e12d`
- js is also duplicated:
Screenshot 2024-05-21 at 14 02 15

Without being sandboxed: When appending `&calypsoify=1` I still see the Fullscreen mode option:

Screenshot 2024-05-21 at 14 45 25

- Script is loaded once, however it looks like styles are not loaded?
Screenshot 2024-05-21 at 14 39 06

Is the desired behaviour for Simple sites the one we see without being sandboxed?

**Comment by coder-karen** [comment]
@fgiannar thanks for highlighting that! That is interesting - when Sandboxed the assets do load, even just on trunk with no additional changes. But when sandboxed, they don't.

What I've noticed is that when sandboxed, admin-plugins is used, but it isn't in production. There is probably some documentation somewhere around that, I can look into. Also, the mu-wpcom changes are the only ones used on Simple sites (not the package required via Jetpack).

As for the duplicated scripts, that was because of the package being loaded via mu-wpcom plugin as well as the admin-plugin. 

In short, this is what I think I should do:
- Make any CSS tweaks that affect Simple directly in admin-plugins.
- Make sure the mu-plugins package version is not loaded on Simple.

**Comment by coder-karen** [comment]
I've edited the PR description with more info and removed Simple testing steps - I did add the CSS tweak directly to the admin-plugins version as well, in the linked diff. I realized that on Simple the module from Jetpack was not being loaded at all previously, so it left two options: Either remove `admin-plugins/calypsoify`, or ensure the package version is only loaded on Atomic sites. To keep things more in parity with how they were I've opted for leaving the admin-plugins version and made sure the mu-wpcom package only initialized the Calypsoify class if a site is Atomic. 

There are also some differences between the two - JS relevant only to Simple in the admin-plugins version, and some minor PHP relevant to Simple too. I imagine it could probably be done (moving to mu-wpcom package version), but that would also require even more research to see what other side effects there may be. I think initializing the package version only in Atomic makes sense just now.

File: projects/packages/calypsoify/.phan/baseline.php [hunk 12]
```diff
@@ -9,12 +9,11 @@
  */
 return [
     // # Issue statistics:
-    // PhanTypeMismatchArgumentProbablyReal : 1 occurrence
     // PhanTypeMismatchPropertyDefault : 1 occurrence
 
     // Currently, file_suppressions and directory_suppressions are the only supported suppressions
     'file_suppressions' => [
-        'src/class-jetpack-calypsoify.php' => ['PhanTypeMismatchArgumentProbablyReal', 'PhanTypeMismatchPropertyDefault'],
+        'src/class-jetpack-calypsoify.php' => ['PhanTypeMismatchPropertyDefault'],
```
Comment (fgiannar) [inline projects/packages/calypsoify/.phan/baseline.php]: I believe we can get rid of this entire file by fixing the issue in `class-jetpack-calypsoify.php` line 23 by replacing with `@var object` with `@var self|false`

File: projects/packages/calypsoify/src/class-jetpack-calypsoify.php [hunk 13]
```diff
@@ -33,7 +34,7 @@ class Jetpack_Calypsoify {
 	/**
 	 * Jetpack_Calypsoify constructor.
 	 */
-	private function __construct() {
+	public function __construct() {
```
Comment (fgiannar) [inline projects/packages/calypsoify/src/class-jetpack-calypsoify.php]: Since we are still using the Singleton pattern to load this class, why are we changing this to public?

Also there's a `Jetpack_Calypsoify::get_instance();` in line `221`. Is this intended?

**Review by fgiannar — COMMENTED** [review]
Great work, Karen! Many thanks!

Left a few inline comments. Other than that I tested the functionality on self-hosted, WoA and Simple and everything works as described 👍

File: projects/packages/calypsoify/.gitattributes [hunk 14]
```diff
@@ -14,4 +14,4 @@ changelog/**      production-exclude
 phpunit.xml.dist  production-exclude
 .phpcs.dir.xml    production-exclude
 tests/**          production-exclude
-.phpcsignore      production-exclude
+.phpcsignore      production-exclude
```
Comment (fgiannar) [inline projects/packages/calypsoify/.gitattributes]: Needs a new line at EOF :)

File: projects/packages/calypsoify/.phan/baseline.php (hunk 12 above)
Comment (coder-karen) [reply projects/packages/calypsoify/.phan/baseline.php]: Nice, thanks!

File: projects/packages/calypsoify/src/class-jetpack-calypsoify.php (hunk 13 above)
Comment (coder-karen) [reply projects/packages/calypsoify/src/class-jetpack-calypsoify.php]: > Since we are still using the Singleton pattern to load this class, why are we changing this to public?

No good reason that I can think of just now, I may have thought it necessary when initially moving code over. Removed now.

> Also there's a Jetpack_Calypsoify::get_instance(); in line 221. Is this intended?

Ah good catch. Since we're calling `get_instance` in both the mu-wpcom package and the Jetpack module directly, that'll be a no. Removed now.

**Review by fgiannar — DISMISSED** [review]
LGTM 👍

File: projects/packages/calypsoify/src/scss/style-gutenberg.scss [hunk 15]
```diff
@@ -5,7 +5,8 @@
  * Hide the last item of the first group when the screen is big enough.
  */
 @media (min-width: 782px) {
-  .components-dropdown-menu__menu:first-child .components-menu-group:first-child .components-button:last-of-type {
+  .interface-more-menu-dropdown__content .components-dropdown-menu__menu:first-child .components-menu-group:first-child .components-button:last-of-type,
+  .more-menu-dropdown__content .components-dropdown-menu__menu:first-child .components-menu-group:first-child .components-button:last-of-type {
```
Comment (coder-karen) [inline projects/packages/calypsoify/src/scss/style-gutenberg.scss:9]: This prevents an issue with other dropdown menu items being hidden when Calypsoified (for example 'Mobile' when clicking the Preview icon). See D149786-code for more context.
Two different selectors are used because `.interface-more-menu-dropdown__content` is the correct classname in WordPress core (also used in Jetpack), but `.more-menu-dropdown__content` is what is used on Atomic and Simple. Keeping the core version for additional safety.

**Review by fgiannar — APPROVED** [review]
Thanks for the CSS hotfix 👍

**Comment by coder-karen** [comment]
Flagging this Slack thread - p1716540893951409/1716452544.422759-slack-C05PV073SG3 - some testing discrepancies noticed this morning related to the WordPress.com post editor URLs (WordPress.com vs wp-admin URLs from the WordPress.com posts page if this branch is checked out, vs not). It looks as though this was localized to existing test sites, not replicable elsewhere, but noting it here too.
